### PR TITLE
Add a new command `azd binding` to manage service bindings 

### PR DESCRIPTION
**Pull request by houk-ms**
### Command Interface
- `azd binding`: Create bindings for all services in the current project.
- `azd binding api`: Create bindings for the service named 'api'.

Check [azd-binding.snap](https://github.com/Azure/azure-dev/pull/3549/files#diff-6179b6aea40b75469bca353cb5610a7f6e004e1a2e540d19e02c6376c0b6a040) for more details.

### Command Behavior
- Take the `bindings` property in the `azure.yaml` file, and create bindings for the corresponding service. A [service connector resource](https://learn.microsoft.com/en-us/azure/service-connector/overview) is created hehind each binding.
- `azd binding` should be as a step of `azd up`, it runs after `azd provision` and before `azd deploy` 
- `azd binding` is an alpha feature

### User Interface
To use the service binding experience, the `azure.yaml` file, `infra/main.bicep` file and application codes need work together. 

- Update `azure.yaml` file to add bindings for a service, an example as following. Check [schemas/alpha/azure.yaml.json](https://github.com/Azure/azure-dev/compare/main...houk-ms:bindings?expand=1#diff-53fae8f160e3407633aea560b547832aa84ec0f0bc5d5541b29b078789d97967) for a full schema of `azd binding` supported properties.
```yaml
services:
  web:
    project: ./src/web
    language: js
    host: containerapp
    bindings:
      - name: web2api
        targetType: containerapp
        targetResource: api    
  api:
    project: ./src/api
    language: python
    host: containerapp
    bindings:
      - name: api2cosmos
        targetType: cosmos
        targetResource: todocosmos
```

- Define necessary bicep outputs (azd envs) for the bindings to consume. For example, `azd binding` looks for the real resource name of `todocosmos` in `.env` file by the key `BINDING_RESOURCE_TODOCOSMOS`, so it should be an output in `infra/main.bicep`. Check [cli/azd/pkg/binding/binding_config.go](https://github.com/Azure/azure-dev/compare/main...houk-ms:bindings?expand=1#diff-3df16300d9c3e216704e3de68d7f90c72da6815451948e818b1262cea20d833d) for more details about `azd binding` expected keys.
```
output BINDING_RESOURCE_TODOCOSMOS=cosmos.outputs.cosmosAccountName
```

- `azd binding` will generate binding info for each binding, and the binding info will be stored in an appconfig user could specify. The binding info are the service connection infomation users could use in their application codes e.g., 
  - the service endpoint when binding a compute service
  - the database connection string when binding a database service

### CODE REVIEW

**Comment by houk-ms** [comment]
Hi @ellismg, I have just completed the first version of `azd binding` command. Your feedback and suggestions are highly appreciated. Please let me know if there are any areas of improvement or if you have any questions about the changes.

File: cli/azd/pkg/binding/binding_config.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/servicelinker/armservicelinker/v2"
+)
+
+// A child property of `services` in azure.yaml, used to define bindings of a service
+type BindingConfig struct {
+	Name           string             `yaml:"name"`
+	TargetType     TargetResourceType `yaml:"targetType"`
+	TargetResource string             `yaml:"targetResource"`
+	StoreType      StoreResourceType  `yaml:"StoreResourceType"`
+	StoreResource  string             `yaml:"storeResource"`
```
Comment (yungezz) [inline cli/azd/pkg/binding/binding_config.go:13]: how to customize key names to consistent with code?

File: cli/azd/pkg/binding/linker_config.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		"configurationStores/%s",
+	TargetTypeServiceBus:  "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.ServiceBus/namespaces/%s",
+	TargetTypeSignalR:     "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.SignalRService/SignalR/%s",
+	TargetTypeWebPubSub:   "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.SignalRService/WebPubSub/%s",
+	TargetTypeAppInsights: "/subscriptions/%s/resourceGroups/%s/providers/microsoft.insights/components/%s",
+	// compute service as target, the format may be different from source resource
+	TargetTypeWebApp:       "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Web/sites/%s",
+	TargetTypeFunctionApp:  "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Web/sites/%s",
+	TargetTypeContainerApp: "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.App/containerApps/%s",
+	TargetTypeSpringApp:    "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.AppPlatform/Spring/%s",
```
Comment (yungezz) [inline cli/azd/pkg/binding/linker_config.go:50]:
```suggestion
	TargetTypeSpringApp:    "/subscriptions/%s/resourceGroups/%s/providers/Microsoft.AppPlatform/Spring/%s/apps/%s",
```

File: cli/azd/pkg/binding/linker_manager.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	linkerConfig *LinkerConfig,
+) armservicelinker.LinkerResource {
+	// Fixed to use secret as auth type for azd
+	secretAuthType := armservicelinker.AuthTypeSecret
+	azureResourceType := armservicelinker.TargetServiceTypeAzureResource
+	secretTypeRawValue := armservicelinker.SecretTypeRawValue
+
+	return armservicelinker.LinkerResource{
+		Properties: &armservicelinker.LinkerProperties{
+			AuthInfo: &armservicelinker.SecretAuthInfo{
```
Comment (yungezz) [inline cli/azd/pkg/binding/linker_manager.go:141]: opt-out auth and network?

**Comment by ellismg** [comment]
Thanks, for starting this conversation @houk-ms.  I'll take a look shortly!

File: cli/azd/pkg/binding/linker_manager.go (hunk 3 above)
Comment (houk-ms) [reply cli/azd/pkg/binding/linker_manager.go:141]: network should be opted-out, auth should be not. will update

File: cli/azd/pkg/binding/binding_config.go (hunk 1 above)
Comment (houk-ms) [reply cli/azd/pkg/binding/binding_config.go:13]: to keep a simple experience, we may not add it in this pr. 
while we will track it as a todo feature, and support it if users have requests in the future

File: cli/azd/pkg/binding/linker_config.go (hunk 2 above)
Comment (houk-ms) [reply cli/azd/pkg/binding/linker_config.go:50]: will remove the compute services as targets except containerapp, to align with service connector backend support

**Comment by houk-ms** [comment]
Hi @ellismg, could you please help review the PR ?

By the way, i have no idea why the CI build fails with error `Access denied. internal Build Service (azure-sdk) needs Edit build quality permissions for vstfs:///Build/Build/3700208 in team project internal to perform the action.` Any ideas?

**Pull request closed without merging**